### PR TITLE
add stuff

### PR DESCRIPTION
**Pull request by vincanger**

### CODE REVIEW

File: .config/installer.sh [hunk 1]
```diff
@@ -0,0 +1,270 @@
+#!/bin/sh
```
Comment (Martinsos) [inline .config/installer.sh:1]: Just a reminder to what we discussed: let's asap remove this from here and pull it from ze internet! That means that version of installer that we host online will need to support choosing destination for where to install Wasp, so we need to do a PR for it that will add that feature. But that is great, we wanted to add that feature anyway.

File: .config/installer.sh [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    URL="https://app.posthog.com/capture"
+    HEADER="Content-Type: application/json"
+
+    if [ -z "$WASP_TELEMETRY_DISABLE" ]; then
+        if has_curl; then
+            curl -sfL -d "$DATA" --header "$HEADER" "$URL" > /dev/null 2>&1
+        elif has_wget; then
+            wget -q --post-data="$DATA" --header="$HEADER" "$URL" > /dev/null 2>&1
+        fi
+    fi
```
Comment (Martinsos) [inline .config/installer.sh:259]: I also wonder what about this telemetry here. We don't want these "installs" to count as normal installs, I think. Instead we want them to not count at all, or count as special installs, in the sense they are flagged as replit installs. Btw what are we doing on Gitpod? We should be doing the same thing both there and here. I wonder now if we are handling this on Gitpod or not.

File: .config/installer.sh (hunk 1 above)
Comment (Martinsos) [inline .config/installer.sh:1]: Commenting here because I can't comment on binary files -> .DS_Store and .config/.DS_Store -> we don't really want to track those, right? If so pls gitignore them.

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,5 @@
+/.wasp/
+/.env.server
+/.env.client
+/node_modules
+/.*[a-zA-Z0-9]{15,}
```
Comment (Martinsos) [inline .gitignore]: Oh wow I didn't know gitignore supports this kind of regexes!

What does this regex stand for? any amount of chars in front, and then 15 or more alphanums? What is that for?

File: .replit [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+[nix]
+channel = "unstable"
+
+[env]
+XDG_CONFIG_HOME = "/home/runner/.config"
+PATH = "/home/runner/$REPL_SLUG/.config/bin:$PATH"
+REACT_APP_API_URL="https://$REPL_SLUG.$REPL_OWNER.repl.co:8099"
+WASP_WEB_CLIENT_URL="https://$REPL_SLUG.$REPL_OWNER.repl.co:9000"
+WASP_TELEMETRY_DISABLE="1"
+WASP_TELEMETRY_CONTEXT="replit"
```
Comment (Martinsos) [inline .replit:19]: Is it possible to put comments in this file format? If so, could you put some, explaining why we put certain value to what we put them to?

File: .replit [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+PRISMA_CLIENT_ENGINE_TYPE="binary"
+PRISMA_CLI_QUERY_ENGINE_TYPE="binary"
+
+[languages.wasp]
+pattern = "**/*..wasp"
+
+[languages.wasp.languageServer]
+start = "wasp waspls"
```
Comment (Martinsos) [inline .replit:31]: This is cool!

File: README.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
+🚨 IMPORTANT! You must visit your repl's url at port 9000 to correctly preview your app: </br>
```
Comment (Martinsos) [inline README.md:10]: And this is not normal?

File: package.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (Martinsos) [inline package.json]: This file, and also package-lock.json, sound redundant?

File: replit.nix [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{ pkgs }: {
+	deps = [
+		pkgs.lsb-release
+    pkgs.nodejs-18_x
```
Comment (Martinsos) [inline replit.nix]: Weird indentation?

File: run.sh [hunk 9]
```diff
@@ -0,0 +1,3 @@
+sed -i "3c\
+https://${REPL_SLUG}.${REPL_OWNER}.repl.co:9000" README.md
```
Comment (Martinsos) [inline run.sh]: Oh, this inserts correct URL to open the app into README.md?
In that case, I would make it more robust! Right now, I can hapilly modify the README to improve it and unknowingly break this.
So what I would suggest is that you don't rely on just replacing whatever is in the line number 3, but instead putting a special string in the README.md and then sed finds and replace that string. You could put something like <APP_URL_WILL_GET_AUTOMATICALLY_INSERTED_HERE> and search for that -> you can use string that makes more sense, but this is the idea.
Then if I am editing README.md, I can't mess it up by just adding stuff around, only way to mess it up is to edit that string, which I am very unlikely to do since it is obviously a placeholder for some automatic process.

File: main.wasp [hunk 10]
```diff
@@ -0,0 +1,63 @@
+app WaspReplit {
```
Comment (Martinsos) [inline main.wasp:1]: Should we scope the whole wasp.app one more level? Into something like `app/`
So it would be:

```
replit.nix
...
- app
  - main.wasp
  - src/
    - ..
```

I am thinking it might be useful to make it clear what is Wasp app, and has it eh rest of the stuff (well, replit stuff).
But, if you think that makes it a bit less accessible or harder to work with, I understand.

**Review by Martinsos — COMMENTED** [review]
@vincanger nice job, and congrats on tackling all the Replit stuff :)!

All looking very good, mostly what is left is some reorganization, polish, comments, ...  -> check out my comments.

Telemetry wise -> Wasp stores some data about project (its "fingerprint") in tmp location, which enables telemetry to:
1. not send itself too often (because it stores when it was sent last time)
2. to identify project as the same project from before.

This can be tricky to maintain on platforms like gitpod/replit because they often clean up the state on disk between every run / restart. On gitpod we used information we have on the gitpod workspace / gitpod user / something like that to explicitly, based on that, identify user / project in our telemetry, which mostly fixes the problem.

What about replit? What is the lifecycle of the replit "app" (or what is it called)? Does it save data we have in tmp dirs? We need to figure this out, and make sure that we are tracking a specific user/project as correctly as possible, or we will think there are much more unique users and project than there are.

File: .config/installer.sh (hunk 1 above)
Comment (vincanger) [reply .config/installer.sh:1]: This brings up a good point. Replit allows users to create templates, so we could e.g. create a template which users can fork, and Wasp and the ToDo app will already be installed and ready to run. This would of course mean that the template would need to be updated whenever a new Wasp version comes out. I'm not sure at the moment how that would work. I'm hoping that a published template can be updated from time to time without having to republish. This would allow users to demo Wasp right away and avoid having to install wasp and compile an app from the beginning.

File: .config/installer.sh (hunk 2 above)
Comment (vincanger) [reply .config/installer.sh:259]: Gitpod uses the normal installer from ze internet

File: .config/installer.sh (hunk 1 above)
Comment (vincanger) [reply .config/installer.sh:1]: 👍

File: .gitignore (hunk 3 above)
Comment (vincanger) [reply .gitignore]: I don't think it actually worked, tbh. But yes, replit was creating hidden temp files that were getting pushed to github and I was trying to avoid them

File: .replit (hunk 4 above)
Comment (vincanger) [reply .replit:19]: good idea

File: README.md (hunk 6 above)
Comment (vincanger) [reply README.md:10]: Correct. I think, because Wasp is exposing two apps that communicate via different ports, that the default url that replit exposes the client on (or whtever is running on port 3000) was breaking. That's why you guys were seeing the 'REPL WAKING UP' message. When exposing it on a different external port, I was able to avoid that error

File: run.sh (hunk 9 above)
Comment (vincanger) [reply run.sh]: Very good idea, I hadn't thought of that

File: main.wasp (hunk 10 above)
Comment (vincanger) [reply main.wasp:1]: Yeah I avoided that because I hid all the other files and directories that aren't wasp related. See the `hidden` property in the .replit fle

**Comment by vincanger** [comment]
> What about replit? What is the lifecycle of the replit "app" (or what is it called)? Does it save data we have in tmp dirs? We need to figure this out, and make sure that we are tracking a specific user/project as correctly as possible, or we will think there are much more unique users and project than there are.

@Martinsos  -- Files persist on Replit as long as it is not saved in `$HOME./local` within the REPL's filesystem, I believe.  I'm not sure how to investigate that at the moment. I guess at this point, with an instantiated Wasp app on replit and with WASP_TELEMETRY_CONTEXT="replit", I could turn on WASP_TELEMETRY and poke around right?

File: .config/installer.sh (hunk 1 above)
Comment (Martinsos) [reply .config/installer.sh:1]: Ok interesting, let me think through it for a bit!

So right now, we have following steps:
1. Run installer (right now it is local, but normally we would pull it via curl). Installer downloads Wasp and installs it on the user's machine. That basically means putting a couple of files in correct places and that is it.
2. We generate Wasp source files to get started with Wasp app. Right now, we do that by already having files ready in the template. It could also be done with `wasp new <ProjectName>`.
3. User then runs `wasp start`, which smartly recognizes that node deps that the generated wasp code uses were never installed and therefore installs them (which takes a minute). Then, wasp app actually runs!

So, stuff that can be prepared up front:
1. Wasp binary / installation -> we could just put those in place where they need to be and that is that. Yes, that means it is a specific version of Wasp, and for the different version we need to update template. We actually don't gain much with this, since installation is quick anyway.
2. I think the way we have prepared Wasp source files right now is good. They are also version specific though! If new wasp version comes, with breaking changes, it is likely these files will stop working. Alternative is to generate starting project with `wasp new <MyProject>` -> that would ensure it is aligned with whatever version of Wasp we have - but then it will have much less content.
3. Installing node modules is slow, takes a whole minute! Avoiding that would be quite interesting. In theory, we could include those also in the template. They are pretty big though, bigger than anything else. And I am not sure if it will even work, in case some stuff from them is compiled/donwloaded based on the machine we are building them on. Like Prisma binaries. Uff yeah, that doesn't sound right now. Ok, I think this is a no go.

I had this idea with node_modules but now I don't like it any more.

I think we have two directions:
1. Install Wasp (via curl!) and then use a wasp command `wasp new <MyApp>` to generate initial project. This means people get the latest Wasp without modifying template (although we can also pin down the version by providing argument to installer).
Bad thing is that we start with very simple app then, one provided by `wasp new`, instead of the more rich app, like the one you provide here in the template. That could be potentially solved by enriching wasp CLI with smth like `wasp new --<template_name> <MyApp>`, so you could do smth like `wasp new --complex MyApp` and then it kick starts you with a richer app.
2. Come with pre-installed Wasp of specific version and have the Wasp source pre-prepared as you have now. Then we need to change template to use newer version of Wasp.

But maybe it is ok to have to change template for newer version of Wasp? Is that a big issue?

OK to summarize: installation we can do via curl or have stuff preinstalled, it doesn't really matter, so I would probably recommend doing it via curl because it is simpler.
And project we can already have ready, or we can create it.

I am not sure what I like better. Maybe go simple: install via curl, pre-prepared project, version is specific but fine, and there we go.

File: .config/installer.sh (hunk 2 above)
Comment (Martinsos) [reply .config/installer.sh:259]: OK that makes sense, but does it set WASP_TELEMETRY_DISABLE to 1 when running installer? If not, then that means that every running of gitpod counts as instal which is not great and we should improve that and have a better solution both for replit and gitpod.

File: .gitignore (hunk 3 above)
Comment (Martinsos) [reply .gitignore]: Ok then let's remove the regex and let me know if you need help with those tmp files from replit.

File: README.md (hunk 6 above)
Comment (Martinsos) [reply README.md:10]: ha ok interesting. That is because our server is running on 3000 I believe. But in theory, we could make server run on some other port! Than we wouldn't have this problem with Replit. We actually have an issue for that here: https://github.com/wasp-lang/wasp/issues/730 .
But as is mentioned in this issue, you can use ENV vars to do it right now! Yeah, so let's do it maybe? Let's set server to use a different port than 3000, and we can then set frontend to use 3000, if that is what replit likes, and we get the proper behavior and don't have to do any warnings here?

File: main.wasp (hunk 10 above)
Comment (Martinsos) [reply main.wasp:1]: Ok, makes sense!

**Comment by Martinsos** [comment]
> > What about replit? What is the lifecycle of the replit "app" (or what is it called)? Does it save data we have in tmp dirs? We need to figure this out, and make sure that we are tracking a specific user/project as correctly as possible, or we will think there are much more unique users and project than there are.
> 
> @Martinsos -- Files persist on Replit as long as it is not saved in `$HOME./local` within the REPL's filesystem, I believe. I'm not sure how to investigate that at the moment. I guess at this point, with an instantiated Wasp app on replit and with WASP_TELEMETRY_CONTEXT="replit", I could turn on WASP_TELEMETRY and poke around right?

If you enable wasp telemetry by not setting WASP_TELEMETRY_DISABLE, you can run `wasp telemetry` in a project and it will tell you where it keeps its cache.
You can then try to figure out if Replit restarts those files or not, that shouldn't be hard to find out, just check if on next start they are there.

If it turns out these files persist per one Replit project, that is great, because that means that Wasp telemetry will always correctly identify as the same project, and not new project every time this replit is opened.

There is another thing though -> if a replit user runs multiple wasp projects, it will think different users are running those. It would be great if we could recognize those as a single user. I believe we can tell explicitly wasp telemetry what is the user's id, via ENV var. I think we are doing it in Gitpod. If we can get some kind of Replit ID for user, then we could provide that as user ID to wasp's telemetry and we have solved that.

Yeah, feel free to enable telemetry and poke around. Wasp will not know it is us testing, but will at least know it is replit.
Hmm you could in theory set context to "replit test" while doing this, then later we can use this "test" piece of info to recognize we were testing something. But ok, not a biggie.

**Comment by vincanger** [comment]
@Martinsos  I checked and the telemetry file persisted after multiple refresh/restarts :)

I also added the context and user_id for telemetry

File: .config/installer.sh (hunk 1 above)
Comment (Martinsos) [reply .config/installer.sh:1]: @vincanger what are you going with here at the end?

File: .config/installer.sh (hunk 2 above)
Comment (Martinsos) [reply .config/installer.sh:259]: Bump on this one

File: .replit [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+# Wasp runs a React client & a NodeJS app
+# so we have to provide Wasp with the proper URLs
+# so client & server can communicate with each other
+REACT_APP_API_URL="https://$REPL_SLUG.$REPL_OWNER.repl.co:8099"
+WASP_WEB_CLIENT_URL="https://$REPL_SLUG.$REPL_OWNER.repl.co:9000"
+# Don't remove these values! They ensure that 
+# Prisma and Telemetry work correctly on Replit
+WASP_TELEMETRY_CONTEXT="replit"
+WASP_TELEMETRY_USER_ID="$REPL_OWNER"
+PRISMA_CLIENT_ENGINE_TYPE="binary"
```
Comment (Martinsos) [inline .replit:24]: I think it would be good to comment here on why are we setting client engine type to binary, at least to give some idea of our intentions to the reader of the code.

**Comment by Martinsos** [comment]
@vincanger I gave another look, you improved a lot of things and I resolved those comments, but there are still some comments that are standing out unsolved, pls check those out!

File: .config/installer.sh (hunk 1 above)
Comment (vincanger) [reply .config/installer.sh:1]: I left the template "installed" rather than doing it via curl, as I wanted initial REPL startup time to be as quick as possible.

File: .config/installer.sh (hunk 2 above)
Comment (vincanger) [reply .config/installer.sh:259]: I'm not sure about this one. I didn't change this part of the installer script for Replit, only the install location. So it should be the default install behavior. 

I would argue that we should leave the install script as it is, because we are using Gitpod to give users a chance to try wasp without installing it locally. But in order for them to try it, they would normally have to install it. So it gives us a better idea of how many people are trying out Wasp. If Gitpod users are returning to a Wasp project on gitpod (and actually using it as their dev env) then the installation persists, and it won't install again.

Currently on Replit, if a user forks the Template, because it's pre-installed, we wouldn't be getting these metrics. But the telemetry USER changes for each new user.

File: README.md (hunk 6 above)
Comment (vincanger) [reply README.md:10]: Sorry. I don't think I made this very clear @Martinsos

It doesn't seem that there is anything we can do to avoid this from our side. As it stands, Wasp serves the client on 3000 and the server on 3001. This is fine. This issue though is that for Replit to work correctly with Wasp, we have to configure the `.replit` file so that we expose each service on a different "external port". Replit automatically exposes anything on port 3000 to the standard URL, e.g. `https://wasp-replit-template.vincanger.repl.co/`. Previously, with the client exposed at the default URL, and with the server exposed at an external port (e.g. 5999), the app would break and anyone trying to visit the URL would see the 'REPL WAKING UP' message. I was only able to get it working after specifically exposing the client on its own external port (e.g. 9000) as well.

These are topics I could mention to the Replit Team. Perhaps, because external ports is a new feature, there will be a fix for this in the future

File: .config/installer.sh (hunk 1 above)
Comment (Martinsos) [reply .config/installer.sh:1]: Couple of questions:

1. I am looking at the code and I don't actually see installer.sh used anywhere, what am I missing?
2. Are we always installing the latest Wasp version, or some specific Wasp version? If we are always installing the latest version, how will we ensure it keeps working with the prepared project? If we are installing specific version, how are we going to remember to update the template every so and so, what will be the process around that?
3. They anyway have to wait some time for wasp to install all the npm dependencies, right? I believe time it takes for curling installer.sh is negligible to that. But I guess they are done in different stages? If so, you could probably even move installing of Wasp to happen at the same time ./run happens? Maybe I don't understand something here, sorry if that is so and pls explain me the process!

Comment (vincanger) [reply .config/installer.sh:1]: It seems the way Replit works is different from GitPod. So if a user were to fork this Repl, they get it, as is, with Wasp already installed, but configured for their env/username. 

becase I've already installed Wasp manually, and manage the REPL, if anyone forks it, they get that REPL at it's current state. So, yes, the drawback is that we must maintain it to update to new versions. But, I don't see the harm in trying it out this way, as the idea is to simply get people trying out Wasp. If we miss a couple minor versions (so long as there is no major bug fix necessary) I assumed that would be OK.

EDIT: Maybe it would be a good idea for us to test this path out. Try creating a replit account, and then forking this repl so we can test the user journey: https://replit.com/@vincanger/wasp-template

File: .config/installer.sh (hunk 2 above)
Comment (Martinsos) [reply .config/installer.sh:259]: Ok I think I am missing some knowledge about this Replit preinstallation.
What does that mean? So when a user starts new Wasp Replit, what happens -> new Wasp project is created, but Wasp installation is not performed because that already happened before? At which point? It happens once, globally when we publish new Replit template, something like that?

File: README.md (hunk 6 above)
Comment (Martinsos) [reply README.md:10]: Ahhh got it well that is pretty weird! So for FE and BE to communicate, we have to expose BE via "external port" and then tell FE to use that external port, but that is also not enough because if we leave FE at default 3000 Replit now for some reason doesn't pick it up anymore, so we also have to expose FE via another "external port" and only then it works. Waaaat :D. Ok let's ask them!

File: .config/installer.sh (hunk 1 above)
Comment (Martinsos) [reply .config/installer.sh:1]: Ok, so I understand now that not every Replit template uses this. Instead, they are forked from your Replit, where you already did the installation, and they just continue from there!

**Comment by Martinsos** [comment]
Ok, I think everything is good now.

I think one important thing is to document the current state and processes around Wasp replit, so that we can easily pick this up in the future and so that others can contribute.

I suggest following:
1. In README.md, explain that we currently have https://replit.com/@vincanger/wasp-template, which we expect people to fork. Since our template already has Wasp installed, they don't need to do installation -> instead they just get an already specific version of Wasp installed.
This does mean that when we want to release template with never version of Wasp, we will have to modify our template (install newer version of Wasp on it).
Special NOTE: you used custom version of installer (present in this repo) that allows choosing location of where Wasp is installed, since we need to pick a specific location for Replit to work, so that one should also be used in the future if installing new versions of Wasp.
You could also mention that your template brings a starter project which is a simple Todo App, to get people started.
And this repo basically captures all those files that were used to kick start that initial template.

2. Open an issue describing a weird situation we have with ports, what we did to fix it, ... .

I think that is it, with all this it is well documented I think!

**Comment by vincanger** [comment]
issue about replit template ports added here: https://github.com/wasp-lang/wasp/issues/957

**Review by Martinsos — APPROVED** [review]
Looking good Vince, README now gives some context, we have that issue -> I think we are set! Approved, feel free to merge!
